Establish toolchain and language versioning (#4105)
carbon-language/carbon-lang: docs/project/versioning.md

<!-- toc -->

## Table of contents

-   [Overview](#overview)
-   [Major version increments](#major-version-increments)
    -   [Breaking changes](#breaking-changes)
    -   [Exclusions to what constitutes a breaking change](#exclusions-to-what-constitutes-a-breaking-change)
-   [Minor version increments](#minor-version-increments)
-   [Patch version increments](#patch-version-increments)
    -   [Examples:](#examples)
-   [Pre-release versions](#pre-release-versions)
    -   [Release qualification pre-releases: `MAJOR.MINOR.PATCH-rc.N`](#release-qualification-pre-releases-majorminorpatch-rcn)
    -   [Incremental development versions: `MAJOR.MINOR.PATCH-0.{nightly,dev}.N`](#incremental-development-versions-majorminorpatch-0nightlydevn)
        -   [Nightly pre-release versions](#nightly-pre-release-versions)
        -   [Development pre-release versions](#development-pre-release-versions)
-   [Relevant proposal](#relevant-proposal)

<!-- tocstop -->

## Overview

Carbon uses a single versioning scheme across both the language itself and
toolchain, including the standard library, compiler, linker, and all development
tools released by the main Carbon project.

> The scheme conforms to and is closely based on Semantic Versioning (SemVer):

Reviewer: I think it'd be useful to link to https://semver.org/ here, and maybe specify a version number?). (The idea of semantic versioning predates the SemVer spec, and there are other different and incompatible forms of semantic versioning such as [libtool's system](https://www.gnu.org/software/libtool/manual/html_node/Libtool-versioning.html) so it seems good to be specific.)

Author: Done.

-   Carbon versions: `MAJOR.MINOR.PATCH`

> -   `MAJOR` increments on backwards incompatible changes, including a
>     deprecation that might trigger a build-breaking warning.
> -   `MINOR` increments only expected during early development with a major
>     version of `0`.
> -   `PATCH` increments represent bug fixes only.

Reviewer: For major and patch, we say what the increments mean, whereas for minor we don't say what they mean, only that we won't do them much. If we're going to explain how to understand the version numbers rather than deferring to the SemVer spec, it would be good to say what the minor portion of the version means too.

Something like this would capture what I understood from our prior discussion:
```suggestion
-   `MAJOR` increments on backwards incompatible changes, including a
    deprecation that might trigger a build-breaking warning.
-   `MINOR` increments on backwards incompatible changes prior to
    reaching a major version of `1`, and is expected to remain at `0`
    afterwards.
-   `PATCH` increments represent bug fixes only.
```

It might be clearer to swizzle this around:
```suggestion
-   Releases with backwards incompatible changes, including a
    deprecation that might trigger a build-breaking warning,
    increment `MAJOR` after reaching the `1.0` milestone, and
    increment `MINOR` before then.
-   Releases with only backwards compatible changes are not
    expected to happen.
-   Releases containing only bug fixes increment `PATCH`.
```

-   Pre-release suffixes:
    -   `MAJOR.MINOR.PATCH-rc.N`: The N-th potentially viable candidate for a
        release.
    -   `MAJOR.MINOR.PATCH-0.nightly.YYYY.MM.DD`: A nightly incremental
        development build on a particular day during development of that
        version.
    -   `MAJOR.MINOR.PATCH-0.dev`: An interactive, incremental development build
        on a particular day by some developer during development of that

>         version.

Reviewer: Given that we don't have `-alpha` and `-beta` here, it seems that we don't need the `-0.` prefix to ensure that `dev` and `nightly` are ordered before `rc`. Would it make sense to remove the `0.` part? (We could add it back in future if we add back `alpha` / `beta`.)

Author: I'm hesitant to do this... It seems nice to have `nightly` and `dev` be more explicitly separate categories from `rc` or anything else we have here.

Reviewer: OK. I don't have a strong opinion. And maybe relying on the coincidence that `dev` and `nightly` collate before `rc` is too "clever".

See the sections below for the details of each aspect of these versions.

## Major version increments

Aligned with SemVer, the major version must increment for any _breaking change_
to the language or any part of the toolchain.

The first increment from 0 to 1 is expected to be based on achieving a desired
milestone of feature completeness and quality to declare the language to have
reached a stable version.

Subsequent increments are expected to be done with a time-based release
strategy. Features (or breaking changes) ready to ship will do so, and others
will wait for the next major release. The exact cadence used is future work and

> should be determined based on discussions with customers in the ramp up to and

Reviewer: Is there a better word than "customers" here, that doesn't suggest a financial relationship? Something like "stakeholders" or "affected parties" perhaps?

I'm worried that people might interpret this as our valuing the views of people who give us money over the rest of the community.

Author: Yeah, "users" is a more general and better term.

At some future point, it is possible that Carbon will become so stable that it
makes sense to consider using minor version increments for some releases. If and
when this happens, we should revisit our versioning and release policies to
establish a predictable and unsurprising structure.

### Breaking changes

Beyond traditional breaking API changes in either standard library APIs or tool
APIs, Carbon also includes breaking changes in the language or toolchain.
Language and toolchain breaking changes are any that cause correct, functioning,
and non-reflective code to become invalid, rejected, incorrect, or silently
change behavior.

### Exclusions to what constitutes a breaking change

Carbon excludes "reflective code" which is in some way detecting the Carbon
version, or the presence or absence of features and as a consequence can be
"broken" by detecting changes that are correctly designed to otherwise be
non-breaking. We don't want adding features to be considered a breaking change
and so exclude code that specifically detects such additions.

Also excludes breaking changes to incorrect code unless it was accepted and

> functioning in some useful, and typically widespread, way despite its bugs.

Reviewer: This sentence is missing a subject. "Our definition also excludes [...]", or "Carbon also excludes [...]" maybe?

Author: Done.

## Patch version increments

The patch version will increment when the change is fundamentally a bug fix to a
previously released version. We expect the vast majority of these to be strictly
backwards compatible bug fixes.

Patch releases are expected to be driven by demand, and not necessarily present
if unnecessary. However, the exact schedule and process will be determined as
part of getting ready for the 1.0 milestone. Before that milestone we don't
commit to any particular process or cadence for patch releases as no release
before that point should be considered stable.

Note that we still consider restoring the _intended_ "public API" of a release
to be a bug fix. When these bug fixes theoretically break new code in a way that
would typically require a major version increment, they may be made with merely
a patch version increment when they are in fact restoring our intended behavior
for that release. However, we take the SemVer guarantees very seriously as these
fixes can still be disruptive and so we work to hold a high bar for them:

-   They must be in some way fixing a _regressions_ for users from a previous
    release, not merely a missing feature.

>     -   Can even be a regression the overall cohesion or reliability of the

Reviewer: ```suggestion
    -   Can even be a regression in the overall cohesion or reliability of the
```

contributors to Carbon itself to track incremental development progress. That
development-oriented goal drives how they are built and what they do and don't
provide.

Mechanically, we prefix the `nightly` pre-release version component with a `0`
component to ensure these versions sort before any and all release qualification
pre-release versions. We also add a date-derived suffix to provide a rough
chronological ordering of nightly builds with otherwise identical versions.

#### Development pre-release versions

During development, interactive builds of Carbon need to be versioned in an
unambiguous way, and we do that with the `dev` pre-release version. Much like
nightly versions, these are only produced as artifacts of development activities
and are never part of any actual release process. These versions are further not
built expected to be automated or necessarily repeatable. They may contain
in-flight edits and all manner of other variations.

Mechanically, we prefix the `dev` pre-release version component with `0` in the
same way as `nightly` is prefixed to ensure effective ordering. We don't add any
additional information to keep the mechanics of development builds simple --

> there is no easy way to extract the date of the build . The exact timestamp of

Reviewer: Maybe:
```suggestion
for example, there is no easy way to extract the date of the build. The exact timestamp of
```
(but at least remove the space before the period).
